### PR TITLE
Feature

### PR DESCRIPTION
**Pull request by peterrobert**
In this milestone, I completed the following:

  1. Created all my enumerable methods using ruby.
  2. Installed rubocop.
  3. Validated my code and improved it.
  4. Created the readme file.

### CODE REVIEW

File: main.rb [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+      end
+      results
+    else
+      item.my_each do |value|
+        results = value == parms
+        break unless results
+      end
+      results
+    end
+  end
```
Comment (shshamim63) [inline main.rb]: **`my_all?`** identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2)) 
- Returns true if the block never returns false or nil
```ruby
[1, 2, 3].my_all? { |num| num < 4}   #should return true
```
- when no block or argument is given, returns true when none of the collection members are false or nil
```ruby
[1, 2, 3].my_all? #should return true
[1, 2, nil].my_all? #should return false
[1, false, nil].my_all? #should return false
```
- when a class is passed as an argument returns true if all of the collection is a member of such class
```ruby
[1, 2, 3].my_all?(Integer) #should return true
[1, 2, nil].my_all?(Integer) #should return false
```
- when a Regex is passed as an argument returns true if all of the collection matches the Regex
```ruby
%w[dog door rod blade].my_all?(/d/)   #should return true.
```
- when a pattern other than Regex or a Class is given, returns true if all of the collection matches the pattern
```ruby
[3, 3, 3].my_all?(3) #should return true
[3, 3, 1].my_all?(3) #should return false
```

File: main.rb [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+      end
+    elsif item.empty? && parms[0].nil?
+      res = false
+    else
+      item.my_each do |i|
+        res = true if i
+      end
+    end
+    res
+  end
```
Comment (shshamim63) [inline main.rb]: `my_any?` identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- when a class is passed as an argument returns true if at least one of the collection is a member of such class
```ruby
[1, 2, 3].my_any?(Integer) #should return true
[1, 'demo', false].my_ant?(Integer) #should return true
['demo', false, nil].my_any?(Integer) #should return false
```
- when a Regex is passed as an argument returns false if none of the collection matches the Regex
```ruby
%w[dog door rod blade].my_any?(/d/)   #should return true.
```

File: main.rb [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+        return false if yield(item)
+      elsif parms.length == 1
+        return false if parms[0] == item
+      elsif item == true
+        return false
+      end
+      return true if i == length - 1
+    end
+    true if length.zero?
+  end
```
Comment (shshamim63) [inline main.rb]: `my_none?` identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- when a class is passed as an argument returns true if none of the collection is a member of such class
```ruby
[1, 2, 3].my_none?(String)  #should return true
['1', '2', '3'].my_none?(String)  #should return false
```
- when a Regex is passed as an argument returns true only if none of the collection matches the Regex
```ruby
%w[dog door rod blade].my_none?(/z/)   #should return true
```

File: main.rb [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+      parms = results.to_proc
+      results = nil
+    end
+
+    sym = sym.to_sym if sym.is_a?(String)
+    parms = sym.to_proc if sym.is_a?(Symbol)
+
+    my_each { |i| results = results.nil? ? i : parms.yield(results, i) }
+    results
+  end
```
Comment (shshamim63) [inline main.rb:161]: `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- combines all elements of enum by applying a binary operation, specified by a block
```ruby
[1, 2, 3].my_inject { |memo, num| memo + num } #should return 6
[1, 2, 3].my_inject(1) { |memo, num| memo + num } #should return 7
```
- when a symbol is specified combines each element of the collection by applying the symbol as a named method
```ruby
[1, 2, 3].my_inject(:+)  #should return 6
[1, 2, 3].my_inject(1, :+)  #should return 7
```

File: main.rb [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+
+  def multiply_els(my_array)
+    my_array.my_inject { |result, value| result * value }
+  end
```
Comment (shshamim63) [inline main.rb]: `multiply_els` is not an enumerable method. The purpose of it, to check whether you can access your enumerable methods properly or not. So it should be placed outside of the module.

**Review by shshamim63 — CHANGES_REQUESTED** [review]
## Status :space_invader: Required Changes :space_invader: 
___
Hi @peterrobert :wave: ,
Awesome job :clap: . Most of your methods are performing properly. But I have found some methods with corner cases where they are failing. Which is blocking me to approve this project right now. I have added some comments, hope it will help you to fix the issues.
___
### :spades: Mandatory Changes :spades: 
___
Here I will list the methods that are not working properly. Just match them with the comments below to find the instruction.
1. my_all?
2. my_any?
3. my-none?
4. my_inject
5. multiply_els
___
#### Please fix the above-mentioned issues and send another review request.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain
[Linkedin](linkedin.com/in/shakhawathossainshamim/)

**Comment by peterrobert** [comment]
In this Milestone, I have done the following:

1. fixed all suggested changes by the review

File: main.rb [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+      parms = results.to_proc
+      results = nil
+    end
+
+    sym = sym.to_sym if sym.is_a?(String)
+    parms = sym.to_proc if sym.is_a?(Symbol)
+
+    my_each { |i| results = results.nil? ? i : parms.yield(results, i) }
+    results
+  end
```
Comment (euqueme) [inline main.rb:161]: - [x] should have working `my_inject` **(covered by automated tests)** identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

####  my_inject

- [x] combines all elements of enum by applying a binary operation, specified by a block
```ruby
(1..5).my_inject(4) { |prod, n| prod * n } == (1..5).inject(4) { |prod, n| prod * n } => 480
```
- Yours is retrieving a NoMethodError
image

- Real inject
image

- [x] When a symbol is provided as an argument, my_inject should combine the elements of the enum using that symbol as a named method.
```ruby
(1..50).my_inject(:+) == (1..50).inject(:+) => 1275
```
- Yours is retrieving a NoMethodError
image

- Real inject
image

**Review by euqueme — CHANGES_REQUESTED** [review]
> ### Pull Request under the review by Maria Eugenia Queme P. (TSE)

### Hello @peterrobert :handshake: 

### Thank you for doing the changes requested by the previous reviewer :medal_sports: also claps to you for doing step # 12 not all students get it in the first review but you did :clap: :clap: :clap: only one of your enumerable methods is not working properly please address the comments I've left in your code :point_down: 

### That's it! :raised_hands: 

### Please :pray: do the changes :man_technologist: and submit back for a code review when you are done :memo: 

Happy Coding :rocket: 

Any questions? ask me on slack **@ Maru** :woman_technologist: 

PROJECT STATUS: REQUESTED CHANGES :vertical_traffic_light:

**Comment by peterrobert** [comment]
In this Milestone I have completed the following:
1. Fixed the changes requested from the code review. All I needed to do was convert a range to an array at my_each method.

**Review by Torres-ssf — APPROVED** [review]
### Project Approved :tada: :clap: 
Hi there @peterrobert, how are you?
Great job on this assignment and implementing the requested changes.
**Your PR is approved**
Go luck with the next project :+1: 

Happy coding :computer: 
**Sergio Torres**

**Pull request merged**